records/2023-06-27.scalability.increasing.send.rates.md

```diff
@@ -0,0 +1,107 @@
+
+## What happens when we send a lot of emails?
+
+Emails are batch saved into the database. This does not appear to be a bottleneck. Each email is send by its own `deliver_email` task.
+
+[`deliver_email()`](https://github.com/cds-snc/notification-api/blob/main/app/celery/provider_tasks.py#L63)
+- Fetches the notification using the id
+- Calls `send_email_to_provider()`
+- So basically a wrapper for send_email_to_provider()
+
+[`send_email_to_provider()`](https://github.com/cds-snc/notification-api/blob/main/app/delivery/send_to_providers.py#L191)
+- Prepares the email
+- Get the template
+- Fill it out
+- Get any files if they’re attached
+- Calls `send_email()` with the html and any attachments
+
+[`send_email()`](https://github.com/cds-snc/notification-api/blob/main/app/clients/email/aws_ses.py#L33)
+- Creates a MIME email with the html and attachments
+- Calls boto3 `send_raw_email()` with the MIME email
+
+So there are basically 3 areas that can take time
+- `send_email_to_provider()` code before the `send_email()` call
+- `send_email()` code before the boto call
+- the boto call
+
+We have timing metrics for:
+- `deliver_email()` (which should be the same as `send_email_to_provider()`): celery_tasks_deliver_email
+- `send_email()`: celery_clients_ses_send_email
+- `send_raw_email()`:  celery_clients_ses_request-time
+
+
+We uploaded a csv of 50K emails (with one variable and no file attachments) to staging (send to `success@simulator.amazonses.com`) on July 12, 2:15 pm EST and monitored the [scaling dashboard](https://ca-central-1.console.aws.amazon.com/cloudwatch/home?region=ca-central-1#dashboards/dashboard/ScalingTemp?start=2023-07-12T18%3A10%3A00Z&end=2023-07-12T19%3A30%3A00Z&autoRefresh=false).
+
+
+![bulk-tasks activity](./diagrams/2023-06-27.scalability.increasing.send.rates/bulk_tasks-queue.png "bulk-task queue")
+
+This graph shows the deliver_email tasks are added to the bulk-tasks queue at a rate of about 7500 / minute. They are being processed at a rate of about 750 / minute.
+
+
+![deliver_email time](./diagrams/2023-06-27.scalability.increasing.send.rates/deliver_email-average-time.png "deliver_email average time")
+
+This graph shows that each deliver_email task took approximately 0.3 seconds during the load test.
+
+
+![deliver_email time breakdown](./diagrams/2023-06-27.scalability.increasing.send.rates/deliver_email-breakdown.png "deliver_email breakdown")
+
+From this graph we see that during the deliver_email task:
+- 65% of the time is in deliver_email code before calling send_email
+- 33% of the time is spent doing the boto call
+- 2% of the time is spent doing send_email code before the boto call
+
+
+## Options
+
+There are two basic areas that can be changed to increase our send rates:
+- Notify code
+- Notify infrastructure
+
+### Notify code
+
+We can try to speed up:
+- the `deliver_email` code before the `send_email` call
+- the boto call (can it be batched somehow?)
```

Comment:
Here is an AWS feature that we discussed very briefly in slack awhile back: https://docs.aws.amazon.com/ses/latest/APIReference-V2/API_SendBulkEmail.html